Compiler/NFFrontEnd/NFSimplifyExp.mo

```diff
@@ -0,0 +1,122 @@
+/*
+ * This file is part of OpenModelica.
+ *
+ * Copyright (c) 1998-2014, Open Source Modelica Consortium (OSMC),
+ * c/o Linköpings universitet, Department of Computer and Information Science,
+ * SE-58183 Linköping, Sweden.
+ *
+ * All rights reserved.
+ *
+ * THIS PROGRAM IS PROVIDED UNDER THE TERMS OF GPL VERSION 3 LICENSE OR
+ * THIS OSMC PUBLIC LICENSE (OSMC-PL) VERSION 1.2.
+ * ANY USE, REPRODUCTION OR DISTRIBUTION OF THIS PROGRAM CONSTITUTES
+ * RECIPIENT'S ACCEPTANCE OF THE OSMC PUBLIC LICENSE OR THE GPL VERSION 3,
+ * ACCORDING TO RECIPIENTS CHOICE.
+ *
+ * The OpenModelica software and the Open Source Modelica
+ * Consortium (OSMC) Public License (OSMC-PL) are obtained
+ * from OSMC, either from the above address,
+ * from the URLs: http://www.ida.liu.se/projects/OpenModelica or
+ * http://www.openmodelica.org, and in the OpenModelica distribution.
+ * GNU version 3 is obtained from: http://www.gnu.org/copyleft/gpl.html.
+ *
+ * This program is distributed WITHOUT ANY WARRANTY; without
+ * even the implied warranty of  MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE, EXCEPT AS EXPRESSLY SET FORTH
+ * IN THE BY RECIPIENT SELECTED SUBSIDIARY LICENSE CONDITIONS OF OSMC-PL.
+ *
+ * See the full OSMC Public License conditions for more details.
+ *
+ */
+
+encapsulated package NFSimplifyExp
+
+import NFExpression.Expression;
+import Operator = NFOperator;
+import Type = NFType;
+
+function simplifyExp
+  input output Expression exp;
+algorithm
+  exp := preSimplify(exp);
+  exp := postSimplify(exp);
+end simplifyExp;
+
+protected function preSimplify
+  input output Expression exp;
+protected
+  Expression exp1, exp2;
+  Operator operator;
+algorithm
+  exp := match exp
+    case Expression.BINARY(exp1=exp1, operator=operator, exp2=exp2)
+      algorithm
+        exp1 := simplifyExp(exp1);
+        exp2 := simplifyExp(exp2);
+      then Expression.BINARY(exp1, operator, exp2);
+
+    case Expression.UNARY(operator=operator, exp=exp1)
+      algorithm
+        exp1 := simplifyExp(exp1);
+      then Expression.UNARY(operator, exp1);
+
+    case Expression.LBINARY(exp1=exp1, operator=operator, exp2=exp2)
+      algorithm
+        exp1 := simplifyExp(exp1);
+        exp2 := simplifyExp(exp2);
+      then Expression.LBINARY(exp1, operator, exp2);
+
+    case Expression.LUNARY(operator=operator, exp=exp1)
+      algorithm
+        exp1 := simplifyExp(exp1);
+      then Expression.LUNARY(operator, exp1);
+
+    case Expression.RELATION(exp1=exp1, operator=operator, exp2=exp2)
+      algorithm
+        exp1 := simplifyExp(exp1);
+        exp2 := simplifyExp(exp2);
+      then Expression.RELATION(exp1, operator, exp2);
+
+    else exp;
+  end match;
+end preSimplify;
+
+protected function postSimplify
+  input output Expression exp;
+protected
+  Expression exp1, exp2;
+  Type ty;
+  Integer i1, i2;
+  Real r1, r2;
+algorithm
+  exp := match exp
```

Comment:
This would probably be faster if rewritten to avoid matching the same thing over and over, like this:
```Modelica
  case Expression.BINARY(exp1 = Expression.INTEGER(value = i1), exp2 = Expression.INTEGER(value = i2))
    then match exp.operator
      case Operator.ADD() then Expression.INTEGER(i1 + i2);
      case Operator.SUB() then Expression.INTEGER(i1 - i2);
      ...
```
I don't think the guards are necessary other than as a sanity check for the typing, but that's not really the simplification's job in my opinion.